src/goose/toolkit/developer.py

```diff
@@ -159,16 +163,163 @@ def shell(self, command: str) -> str:
+        stdout_queue = queue.Queue()
+        stderr_queue = queue.Queue()
+
+        # Function to read stdout and stderr without blocking
+        def reader_thread(pipe: any, output_queue: any) -> None:
+            try:
+                for line in iter(pipe.readline, ""):
+                    output_queue.put(line)
+                    # Check for prompt patterns
+                    for pattern in compiled_patterns:
+                        if pattern.search(line):
+                            output_queue.put("INTERACTION_DETECTED")
+            finally:
+                pipe.close()
+
+        # Start threads to read stdout and stderr
+        stdout_thread = threading.Thread(target=reader_thread, args=(proc.stdout, stdout_queue))
+        stderr_thread = threading.Thread(target=reader_thread, args=(proc.stderr, stderr_queue))
+        stdout_thread.start()
+        stderr_thread.start()
+
+        # Collect output
+        output = ""
+        error = ""
+
+        # Initialize timer and recent lines list
+        last_line_time = time.time()
+        recent_lines = []
+
+        # Continuously read output
+        while True:
+            # Check if process has terminated
+            if proc.poll() is not None:
+                break
+
+            # Process output from stdout
+            try:
+                while True:
+                    line = stdout_queue.get_nowait()
+                    if line == "INTERACTION_DETECTED":
+                        return (
+                            "Command requires interactive input. If unclear, prompt user for required input "
+                            f"or ask to run outside of goose.\nOutput:\n{output}\nError:\n{error}"
+                        )
+
+                    else:
+                        output += line
+                        recent_lines.append(line)
+                        recent_lines = recent_lines[-10:]  # Keep only the last 10 lines
+                        last_line_time = time.time()  # Reset timer
+            except queue.Empty:
+                pass
+
+            # Process output from stderr
+            try:
+                while True:
+                    line = stderr_queue.get_nowait()
+                    if line == "INTERACTION_DETECTED":
+                        return (
+                            "Command requires interactive input. If unclear, prompt user for required input "
+                            f"or ask to run outside of goose.\nOutput:\n{output}\nError:\n{error}"
+                        )
+                    else:
+                        error += line
+                        recent_lines.append(line)
+                        recent_lines = recent_lines[-10:]  # Keep only the last 10 lines
+                        last_line_time = time.time()  # Reset timer
+            except queue.Empty:
+                pass
+
+            # Check if no new lines have been received for 10 seconds
+            if time.time() - last_line_time > 10:
```

Comment:
Curious how this handles cases where log lines are replaced (e.g. print(..., end="\r"). This may be handled differently in stdout vs in a terminal console, but figured I'd float it as something to double check before merging.

Reply:
hrm - might be worth trying out

Reply:
I believe would timeout and then consider if it should continue or not